File: cpp/examples/linux/environmental/BME280/complete/main.cpp
#include <cstdio>
#include <cstdlib>
#include <unistd.h>
#include "I2CConnectionLinux.h"
#include "BME280.h"

int main() {
    const char* bus_env  = getenv("I2C_BUS");
    const char* addr_env = getenv("I2C_ADDR");
    int     bus  = bus_env  ? atoi(bus_env)       : 1;
    uint8_t addr = addr_env ? (uint8_t)strtol(addr_env, nullptr, 0) : 0x76;
    I2CConnectionLinux connection(bus, addr);

    BME280Full bme(connection);                                             // Create BME280 driver, (connection)

    uint8_t id = bme.chip_id();                                            // Read chip ID register, () → uint8_t
                                                                           // BME280 returns 0x60, BMP280 returns 0x58
    printf("chip_id=0x%02X\n", id);

    bme.configure(BME280Full::OSRS_X2, BME280Full::OSRS_X16, BME280Full::OSRS_X1,
                  BME280Full::MODE_NORMAL, BME280Full::FILTER_16, BME280Full::T_SB_0_5_MS);  // Configure in one write, (osrs_t, osrs_p, osrs_h, mode, filter, t_sb) → void
                                                                           // indoor-navigation profile from the datasheet
    bme.set_oversampling(BME280Full::OSRS_X1, BME280Full::OSRS_X1, BME280Full::OSRS_X1);  // Set oversampling, (osrs_t, osrs_p, osrs_h) → void
    bme.set_filter(BME280Full::FILTER_4);                                  // Set IIR filter coefficient, (coeff) → void
                                                                           // smooths pressure against short gusts
    bme.set_standby(BME280Full::T_SB_125_MS);                              // Set normal-mode standby, (t_sb) → void
    bme.set_mode(BME280Full::MODE_NORMAL);                                 // Set power mode, (mode) → void
                                                                           // normal mode measures continuously
    usleep(200000);

    uint8_t st = bme.status();                                             // Read STATUS register, () → uint8_t
                                                                           // STATUS_MEASURING / STATUS_IM_UPDATE bits
    float t = bme.temperature();                                           // Read temperature, () → float °C
    float h = bme.humidity();                                              // Read relative humidity, () → float %RH
    float p = bme.pressure();                                              // Read pressure, () → float hPa
    float alt = bme.altitude(1013.25f);                                    // Estimate altitude, (sea_level_hpa=1013.25 hPa) → float m
                                                                           // barometric formula against the given sea-level pressure
    float slp = bme.sea_level_pressure(100.0f);                            // Reduce to sea level, (altitude_m m) → float hPa
    float dp = bme.dew_point();                                            // Dew point from T and RH, () → float °C
                                                                           // Magnus approximation
    printf("status=0x%02X t=%.2f C h=%.2f %%RH p=%.2f hPa alt=%.1f m slp=%.2f hPa dew=%.2f C\n",
           st, (double)t, (double)h, (double)p, (double)alt, (double)slp, (double)dp);

    bme.reset();                                                           // Soft reset, () → void
                                                                           // all registers return to power-on defaults
    return 0;
}

File: rust/examples/linux/environmental/bme280/complete/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::environmental::{
    Bme280Full, OSRS_X1, OSRS_X2, OSRS_X4,
    MODE_FORCED, FILTER_4, T_SB_125_MS,
};

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x76);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut bme = Bme280Full::new(dev, addr, false).expect("init BME280"); // Create BME280 driver, (i2c, addr=0x76, spi=false)
    let cid = bme.chip_id().expect("read chip id");                 // Read chip ID, () → u8
    println!("chip_id=0x{:02x}", cid);                              // returns 0x60 for BME280
    bme.configure(OSRS_X1, OSRS_X1, OSRS_X1, MODE_FORCED, FILTER_4, T_SB_125_MS).expect("configure");  // Configure chip, (osrs_t 0–5, osrs_p 0–5, osrs_h 0–5, mode 0/1/3, filter 0–4, t_sb 0–7) → ()
                                                                         // writes ctrl_hum, config, ctrl_meas in correct order
    bme.set_oversampling(OSRS_X4, OSRS_X2, OSRS_X1).expect("set osrs");  // Set oversampling, (osrs_t 0–5, osrs_p 0–5, osrs_h 0–5) → ()
                                                                         // humidity update requires ctrl_meas write to latch
    bme.set_mode(MODE_FORCED).expect("set mode");                  // Set power mode, (mode 0/1/3) → ()
    bme.set_filter(FILTER_4).expect("set filter");                 // Set IIR filter, (coeff 0–4) → ()
                                                                         // suppresses short-term pressure disturbances
    bme.set_standby(T_SB_125_MS).expect("set standby");            // Set standby time, (t_sb 0–7) → ()
                                                                         // only relevant in normal mode; codes 6/7 mean 10/20 ms on BME280
    let _st = bme.status().expect("read status");                  // Read status register, () → u8
    let t = bme.temperature().expect("read temperature");          // Read temperature, () → f32 °C
    let p = bme.pressure().expect("read pressure");                // Read pressure, () → f32 hPa
    let h = bme.humidity().expect("read humidity");                // Read humidity, () → f32 %RH
    let alt = bme.altitude(1013.25).expect("compute altitude");    // Compute altitude, (sea_level_hpa=1013.25) → f32 m
                                                                         // uses barometric formula to convert pressure to metres
    let slp = bme.sea_level_pressure(alt).expect("compute slp");   // Compute sea-level pressure, (altitude_m) → f32 hPa
    let dp = bme.dew_point().expect("compute dew point");          // Compute dew point, () → f32 °C
                                                                         // Magnus-Tetens approximation from current T and RH
    bme.reset().expect("soft reset");                              // Soft reset chip, () → ()
                                                                         // re-reads calibration and re-applies configuration
    println!("T={:.1} C, P={:.1} hPa, RH={:.1} %RH, alt={:.1} m, slp={:.1} hPa, dp={:.1} C", t, p, h, alt, slp, dp);
}
